src/harness/simple_crl_server.py

```diff
@@ -0,0 +1,308 @@
+
+
+class SimpleCrlServer(threading.Thread):
+  """Implements a CRL server by creating a HTTP server and
+  serving CRL files The CRL server runs as a separate thread.
+  """
+
+  def __init__(self, crl_url, crl_file_path):
+    """ Constructor for simple CRL Server.
+
+    Args:
+      crl_url: The URL that is mentioned in the CDP extension of the certificates.
+        The CRL server runs on this URL and responds only to this URL with the CRL file.
+      crl_file_path: Absolute path to the CRL file that will be served by this CRL server.
+    """
+    super(SimpleCrlServer, self).__init__()
+    url_components = urlparse(crl_url)
+    self.host_name = url_components.hostname
+    self.port = url_components.port if url_components.port is not None else DEFAULT_CRL_SERVER_PORT
+    self.crl_url_path = url_components.path
+    self.crl_file_path = crl_file_path
+    self.setDaemon(True)
+
+  def run(self):
+    """ Starts the HTTPServer as background thread.
+
+    The run() method is an overridden method from thread class and it gets invoked
+    when the thread is started using thread.start().
+    """
+
+    # Setting the parameters for handler class
+    logging.info("Started CRL Server at %s", self.host_name + ":" + str(self.port))
+    self.server = CrlHttpServer(self.crl_url_path, self.crl_file_path, (self.host_name, self.port),
+                                 CrlServerHttpHandler)
+    self.server.serve_forever()
+
+  def stopServer(self):
+    """This method is used to stop HTTPServer Socket"""
+
+    self.server.shutdown()
+    logging.info('Stopped CRL Server at %s', self.host_name+":" + str(self.port))
+
+
+class CrlHttpServer(HTTPServer):
+  """The class CrlHttpServer overrides built-in HTTPServer and takes the
+  parameter base_path used to serve the files. This is needed to have CRL Server
+  to serve files from different directories.
+  """
+
+  def __init__(self, crl_url_path, crl_file_path, server_address, RequestHandlerClass):
+    self.crl_url_path = crl_url_path
+    self.crl_file_path = crl_file_path
+    HTTPServer.__init__(self, server_address, RequestHandlerClass)
+
+
+class CrlServerHttpHandler(BaseHTTPRequestHandler):
+  """ This class implements the HTTP handlers for the CRL Server. """
+
+  def do_GET(self):
+    """ Handles Pull/GET Request and returns Path of the Request to callback Method. """
+
+    parsed_path = urlparse(self.path)
+    self.log_message('Received GET Request:{}'.format(parsed_path.path))
+
+    # Handles the GET Requests.
+    # If the GET request path does not match the CRL URL path then respond with 404 response.
+    if parsed_path.path != self.server.crl_url_path:
+      self.log_message('Requested url :{0} does not match with CRL Server URL path:{1}'.format(
+        parsed_path.path, self.server.crl_url_path))
+      self.send_response(404)
+      return
+
+    # If the crl_file_path does not exist then respond with 404 response.
+    if not os.path.exists(self.server.crl_file_path):
```

Comment:
There is still a potential filesystem race between `exists()` and `open()`.  Just call `open()` and handle the error.

See https://docs.python.org/2/glossary.html#term-eafp

Reply:
Followed   the EAFP coding style now.